rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,355 @@
+}
+
+```
+
+### Permission Semantics
+
+The precise meaning of individual permissions is left to the database engine.
+The sole exception is the interaction between the `deny` and `allow` parts:
+denied permissions are removed, and the comparison is performed in a
+case-insensitive way after trimming the whitespace. As a special case, `*` is
+allowed as a permission in `deny` part of the role.
+
+For example, if the `allow` permission is `SELECT`, then it can be removed with
+`select`, `SELECT` or `*`.
+
+The engines should prohibit invalid permissions. This is a fatal error and
+should cause connection error.
+
+### Permission Lifecycle
+
+Permissions are applied after the user is automatically provisioned.
+Deprovisioning (deletion/deactivation) of the user should remove ALL permissions
+from the user, without the need to reference the particular user permissions.
+
+The permission synchronization is performed at least once, on connection time:
+
+- database schema is read,
+- import rules are applied,
+- effective permissions are calculated based on user roles,
+- database-side permissions are updated.
+
+Optional, additional syncs may happen as needed (e.g. when a schema change is
+detected).
+
+### TAG Integration
+
+Periodic application of import rules will populate `db_object` resources in the
+backend. These can be imported to TAG, where permission calculation may happen.
+The permission algorithm is based on label matching, which should be a primitive
+operation that is easy to support in TAG. Performance wise it may be necessary
+to filter the set of objects - a single database can viably produce thousands of
+objects - but it should remain a viable approach.
+
+### Backward Compatibility
+
+The new `db_permissions` field should be automatically ignored by Teleport
+versions that don't support it. The new resources will likewise be ignored.
+
+### Audit Events
+
+The [RFD 113](0113-automatic-database-users.md) introduces events
+`db.user.created` and `db.user.disabled`, but these are yet to be implemented.
+
+The `db.user.created` should be extended with an effective list of applied
+permissions. Since the list of database objects may be long, it may be necessary
+to summarize the list of changes in the audit event.
+
+### Observability
+
+The expectation is that permission changes should occur swiftly. If necessary,
+we may consider monitoring the latency of schema queries and the time required
+to apply permissions. This becomes particularly relevant when permissions are
+managed external to the database instance, for instance, through a call to AWS
+Security Token Service (STS).
+
+To enhance observability, it is advisable to introduce appropriate logging.
+Debug-level logs can be used to detail individual permissions granted, while
+keeping in mind that the list might encompass thousands of entries.
+
+### Product Usage
+
+As of the current point in time, there are no plans for the introduction of
```

Comment:
We do need product usage metrics to understand how feature is being used. We should emit Posthog events that tell us how many / which kinds of database objects are being imported, and how many users are actually using this feature (e.g. emit an event every time Teleport does an RBAC check against db_permissions).

Reply:
it can tag along an audit event. for reference: https://github.com/gravitational/cloud/pull/5247/files